### week2 ~ APIs !

<b>that is, _the-web-as-your-filesystem_ ...</b>  &nbsp;&nbsp; (hw2pr1.ipynb)

[the google doc with hw2's details](https://docs.google.com/document/d/1z4HwpUL1-ImGX3j-6bddyZAfdv2SkRQHRNhF5Cu3Fkk/edit?tab=t.0)
<hr>

In [1]:
#
# computing-styling trick of the day     (or, of the minute...)
#
# The setting for word-wrapping on the output is
#     "notebook.output.wordWrap": true,   (in your settings.json file or from Code ... Settings ...) 

print( list(range(10)) )


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


#### Let's see if you already have the requests library...

In [2]:
#
# see if we have the requests library...
#

import requests

In [3]:
#
# If you _don't_ have the requests library, let's install it!
#

# for me, it worked to uncomment and run this command, here in this cell:
# !pip3 install requests  OR   !pip install requests

# an alternative is to run, in a terminal, the command would be 
#  pip3 install requests  OR    pip install requests      (the ! is needed only if inside Python)

# It's very system-dependent how much you have to "restart" in order to use
# the new library (the notebook, VSCode, the Jupyter extension, etc.)

# Troubles?  Let us know!  We'll work on it with you...

In [2]:
#
# hopefully, this now works! (if so, running will succeed silently)
#

import requests

In [4]:
#
# let's try it on a simple webpage
#

#
# we assign the url and obtain the api-call result into result
#    Note that result will be an object that contains many fields (not a simple string)
# 

url = "https://www.cs.hmc.edu/~dodds/demo.html"
result = requests.get(url)
result    

# if it succeeded, you should see <Response [200]>
# See the list of HTTP reponse codes for the full set!

<Response [200]>

[Here is Wikipedia's list of all HTTP response codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
+ 100's: information
+ 200's: success           
+ 300's: redirects
+ 400's + 500's: errors

Perhaps familiar:  404 <br>
Especially fun:  418


In [5]:
#
# when exploring, you'll often obtain an unfamiliar object. 
# Here, we'll ask what type it is 
type(result)

requests.models.Response

How to access the data inside this object, ``result`` ?

One way: [Head over to the online documentation](https://requests.readthedocs.io/en/latest/api/#requests.Response)

In addition, you can "look around" in Python:

In [6]:
# here is one of the data members within the result
# it "remembers" (keeps track of) the url requested:
result.url

'https://www.cs.hmc.edu/~dodds/demo.html'

In [ ]:
# We can print all of the data members in an object with dir
# Since dir returns a list, we will grab that list and loop over it:
all_fields = dir(result)

for field in all_fields:
    if "_" not in field: 
        print(field)

In [8]:
#
# Let's try printing a few of those fields (data members): 
print(f"result.url         is {result.url}")  # the original url
print(f"result.raw         is {result.raw}")  # another object!
print(f"result.encoding    is {result.encoding}")  # utf-8 is very common
print(f"result.status_code is {result.status_code}")  # 200 is success!

result.url         is https://www.cs.hmc.edu/~dodds/demo.html
result.raw         is <urllib3.response.HTTPResponse object at 0x10ae50e80>
result.encoding    is ISO-8859-1
result.status_code is 200


In [ ]:
# In this case, the result is a text file (HTML) Let's see it!
contents = result.text
print(contents)

In [10]:
# Yay!  
# This shows that you are able to "scrape" an arbitrary HTML page... 

# Now, we're off to more _structured_ data-gathering...

#### Traversing the world - and web - <i>without a browser</i>.   

<b>Using the ISS APIs</b> 

+ Here, you'll make some calls using `requests` to, first, the International Space Station API 
+ and, then, the US Geological Survey's earthquake API
+ "API" is short for "Application Programming Interface" 
  + Admittedly, this is not a very informative name:
  + The API is the set of services, which are functions and/or urls, provided by some software or site

Let's try it with the International Space Station api at [http://api.open-notify.org/iss-now.json](http://api.open-notify.org/iss-now.json)
+ [This page has documentation on the ISS API](http://open-notify.org/Open-Notify-API/ISS-Location-Now/)

In [6]:
#
# we assign the url and obtain the api-call result into result
#    Note that result will be an object that contains many fields (not a simple string)
# 

import requests

url = "http://api.open-notify.org/iss-now.json"   # this is sometimes called an "endpoint" ...
result = requests.get(url)
result    

# if it succeeds, you should see <Response [200]>

<Response [200]>

In [7]:
#
# Let's try printing those shorter fields from before:
print(f"result.url         is {result.url}")  # the original url
print(f"result.raw         is {result.raw}")  # another object!
print(f"result.encoding    is {result.encoding}")  # utf-8 is very common
print(f"result.status_code is {result.status_code}")  # 200 is success!

result.url         is http://api.open-notify.org/iss-now.json
result.raw         is <urllib3.response.HTTPResponse object at 0x12511bfd0>
result.encoding    is utf-8
result.status_code is 200


In [ ]:
#
# In this case, we know the result is a JSON file, and we can obtain it that way:

json_contents = result.json()
print(json_contents)

# Remember:  json_contents will be a _dictionary_

{'iss_position': {'latitude': '-42.6203', 'longitude': '-8.7386'}, 'timestamp': 1739069717, 'message': 'success'}


In [22]:
#
# Let's re-remind ourselves how dictionaries work:

json_contents['message']       # Challenge:  could we access the other components? What _types_ are they?!!

'success'

In [24]:
json_contents['timestamp']

1738814153

In [17]:
CS35_Participant_2 = float(json_contents['iss_position']['longitude'])
CS35_Participant_2

-8.7386

In [13]:
#
# In Python, we can use the resulting dictionary... let's see its keys:
print(list(json_contents.keys()))  

# Also, watch out for string vs. numeric types, e.g., for latitude and longitude.
# At heart, _all_ web data are strings... .

# These experiments will be helpful for problem 1, below :)

['iss_position', 'timestamp', 'message']


<font color="Coral"><b>Task 0a</b></font>  &nbsp;&nbsp; From here, 
+ extract the lat and CS35_Participant_2 of the ISS from ``json_contents``  (dictinoary practice!)
+ copy the ``haversine`` function from the assignment
+ find the lat/CS35_Participant_2 of Claremont (or grab it from the hw page! 😊 )
+ combine these to find out how far, in miles, the ISS is from Claremont  
+ <font size="-2">you could _imagine_ writing a function to do all this (no need to - we'll do it with earthquakes!)</font>

In [18]:
lat = float(json_contents['iss_position']['latitude'])
lat

-42.6203

In [30]:
def haversine(lat1, long1, lat2, long2):
    """
    Calculate the great circle distance in kilometers between two points
    on the earth (specified in decimal degrees)
    """
    from math import sin, cos, radians, asin, sqrt   # this import is for the sin, cos, radians, ...
    # convert decimal degrees to radians

    long1, lat1, long2, lat2 = map(radians, [long1, lat1, long2, lat2])

    # haversine formula
    dlong = long2 - long1
    dlat = lat2 - lat1
    trig = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlong/2)**2
    # Radius of earth. Use 3956 for miles. 6371 for km.
    radius = 3956  # we'll use miles!
    return radius * 2 * asin(sqrt(trig))

haversine(-42.6203,-8.7386,34.0967,-117.719)

8651.026294919777

#### Distance between ISS and Claremont
  + 8651.026294919777 miles

## JSON

####  Let's make a brief JSON visit in Python

The library ``json`` allows us to create and read arbitrary JSON data.

In [3]:
# JSON is a javascript dictionary format -- almost the same as a Python dictionary:
data = { 'key':'value',  'fave':42,  'list':[5,6,7,{'mascot':'Aliiien'}] }
print(data)

# we can write in JSON format to a local file, named small42.json:
import json 

with open("small.json", "w") as f:
    json.dump( data, f )

{'key': 'value', 'fave': 42, 'list': [5, 6, 7, {'mascot': 'Aliiien'}]}


In [4]:
# We can also read from a json file
# The resulting data will be a _dictionary_:

with open("small.json", "r") as f:
    dictionary = json.load( f )

print(f"the {dictionary = }")

the dictionary = {'key': 'value', 'fave': 42, 'list': [5, 6, 7, {'mascot': 'Aliiien'}]}


In [15]:
# let's access this dictionary -- first, the keys:
list(dictionary.keys()) # How do we get 'Aliiien' from newdata?

['key', 'fave', 'list']

In [ ]:
# Task: use the dictionary to obtain (a) 'value' , (b) 42 , (c) 'Aliiien'  [tricky!]

# remember that there are two ways to get the value from a key:
# way 1:  dictionary['key']            # errors if 'key' isn't present
# way 2:  dictionary.get('key')        # returns None if 'key' isn't present

dictionary['key'] #a
dictionary['fave'] #b
dictionary['list'][3]['mascot'] #c

'Aliiien'

Most of the time this will be done for us by the ``requests`` library. So, we will simply receive the dictionary of data sent.

Then, the trick is to "extract" the data fragments we want. (Sometimes it feels like forensics - or archaeology!) Try excavating items one **layer** at a time...

#### Remember: &nbsp; <i>not</i> every url returns json data...
+ The url [https://www.cs.hmc.edu/~dodds/demo.html](https://www.cs.hmc.edu/~dodds/demo.html) returns a plain-text file with _markup_ text
+ that is to say, with HTML tags, such as `<title>Title</title>` to designate the components of its content
+ HTML stands for _hypertext markup language_   
+ Often anything with tags similar to `<b>be bold!</b>` is called "markup." 

Let's try our 5C homepages: they're HTML, not JSON:

In [3]:
import requests 

# here, we will obtain plain-text results from a request
url = "https://www.cs.hmc.edu/~dodds/demo.html"  # try it + source
# url = "https://www.scrippscollege.edu/"          # another possible site...
# url = "https://www.pitzer.edu/"                  # another possible site...
# url = "https://www.cmc.edu/"                     # and another!
# url = "https://www.cgu.edu/"
result = requests.get(url)        
print(f"result is {result}")        # hopefully it's 200

result is <Response [200]>


In [ ]:
# if the request was successful, the Response will be [200]. 
# Then, we can grab the text - or json - from the site:

text = result.text                  # provides the HTML page as a large string...
print(f"len(text) is {len(text)}")  # let's see how large the HTML page is... 

print("\nThe first 242 characters are\n")
print(text[:242])                  # we'll print the first few characters...  

# change this to text[:] to see the whole document...
# Notice that we can run many different analyses without having to re-call/re-scrape the page (this is good!)

#### Declined requests!

Caution: <i>See this week's reading!</i>

#### For now let's focus on API calls providing JSON 
+ We're reading-aligned, as we should be!
+ The open-ended problem (the finale in this notebook) offers the _option_ of scraping raw html -- up to you...

<br>

<b>Let's try another ISS "endpoint" ~ one with all of the <i>people</i> in space.</b>

It's at this url:  [http://api.open-notify.org/astros.json](http://api.open-notify.org/astros.json)


In [ ]:
#
# we assign the url and use requests.get to obtain the result into result_astro
#
#    Remember, result_astro will be an object that contains many fields (not a simple string)
# 

import requests

url = "http://api.open-notify.org/astros.json"   # this is sometimes called an "endpoint" ...
result_astro = requests.get(url)
result_astro

# if it succeeded, you should see <Response [200]>

<Response [200]>

In [ ]:
# If the request succeeded, we know the result is a JSON file, and we can obtain it that way.
# Let's call our dictionary something more specific:

astronauts = result_astro.json()
print(astronauts)
d = astronauts     # d is shorter to type

# Remember:  d and astronauts will be a _dictionary_

note = """ here's yesterday's result - it _should_ be the same today!

{"people": [{"craft": "ISS", "name": "Oleg Kononenko"}, {"craft": "ISS", "name": "Nikolai Chub"},
{"craft": "ISS", "name": "Tracy Caldwell Dyson"}, {"craft": "ISS", "name": "Matthew Dominick"},
{"craft": "ISS", "name": "Michael Barratt"}, {"craft": "ISS", "name": "Jeanette Epps"},
{"craft": "ISS", "name": "Alexander Grebenkin"}, {"craft": "ISS", "name": "Butch Wilmore"},
{"craft": "ISS", "name": "Sunita Williams"}, {"craft": "Tiangong", "name": "Li Guangsu"},
{"craft": "Tiangong", "name": "Li Cong"}, {"craft": "Tiangong", "name": "Ye Guangfu"}], "number": 12, "message": "success"}
"""
""

{'people': [{'craft': 'ISS', 'name': 'Oleg Kononenko'}, {'craft': 'ISS', 'name': 'Nikolai Chub'}, {'craft': 'ISS', 'name': 'Tracy Caldwell Dyson'}, {'craft': 'ISS', 'name': 'Matthew Dominick'}, {'craft': 'ISS', 'name': 'Michael Barratt'}, {'craft': 'ISS', 'name': 'Jeanette Epps'}, {'craft': 'ISS', 'name': 'Alexander Grebenkin'}, {'craft': 'ISS', 'name': 'Butch Wilmore'}, {'craft': 'ISS', 'name': 'Sunita Williams'}, {'craft': 'Tiangong', 'name': 'Li Guangsu'}, {'craft': 'Tiangong', 'name': 'Li Cong'}, {'craft': 'Tiangong', 'name': 'Ye Guangfu'}], 'number': 12, 'message': 'success'}


''

This is pretty intricate. Let's try unpacking this - _parsing it_ - with an in-class break-out challenge.

In [ ]:
#
# Cell to try out parsing d  (astronauts)
#

d

Let's compare with a whole other webservice: **earthquakes** 

<br>
<hr>
<br>

#### Earthquake data

[Here is the USGS Earthquate data API documentation](https://earthquake.usgs.gov/fdsnws/event/1/)

Notice that the "headline" is the URL: &nbsp; This is the <i>domain</i> from which we'll access data.
+ Underneath, there are several different <i>endpoints</i> 
+ We are going to focus on the <tt>count</tt> endpoint and the <tt>query</tt> endpoint
+ along with their parameters
  + the whole list of parameters is linked and available by scrolling down
  + that said, it's easy to miss (well, at least I did! :)

First, let's establish that, for these endpoints, we can make requests - by hand - in our browser!

Try this link: <br><br>  [https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01](https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01)

Ok! Let's parse this url. Requests in which there are parameters <i><b>in the url</b></i> are called GET requests:
+ the <b>endpoint</b> is the first part: ``https://earthquake.usgs.gov/fdsnws/event/1/count``
   + Notice that the forward-slashes are very much like our file-system trees from last week!
   + In fact, they usually <i>are</i> file-system trees. They're just on the <i>server</i> side, instead of our "client" side...
+ the <b>parameters</b> follow the question mark: ``format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01``
   + There are four parameters here. Parameters are separated by the ampersand <tt>&amp;</tt> character.
   + Each one is in the format <tt>name=value</tt>  Here are the four:
   + ``format=geojson`` specifies the desired format of the returned data. ``geojson`` is JSON with geographic data.
   + ``minmagnitude=5.0`` specifies the minimum magnitude of earthquakes to consider. 5.0 is strong, if not always catastrophic.
   + ``starttime=2024-01-01`` specifies the earlier time-endpoint to consider. (Jan 1, 2024)
   + ``endtime=2024-02-01`` specifies the later time-endpoint to consider. (Feb 1, 2024)

My result was this:  ``{"count":134,"maxAllowed":20000}``
+ Earthquakes are always happening -- and they do get reclassified. So, the numbers can change - even in the past.

Try it, in your browser.

Also, try increasing the ``minmagnitude`` -- you'll see a progression of fewer and fewer quakes. (Fortunately!) 

<hr>

Then, try it using a short Python script:

In [5]:
#
# Let's try the  count  endpoint, with geojson format (json with geographical data)
#

url = "https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01"

result = requests.get(url)                       # a named input, params, taking the value param_d, above
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # it's nice to be able to see this

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01


In [6]:
# If it worked, we should be able to obtain the JSON. Remember, it's a dictionary. Let's use d:

d = result.json()

print(f"{d =}")

d ={'count': 133, 'maxAllowed': 20000}


#### Handling parameters separately...

It's awkward to include all the parameters as part of the url.

It's much more common to create a <i>dictionary</i> of the parameters, and then pass that to ``requests.get``

Here is an example:

In [7]:
#
# Here is the endpoint
#
url = "https://earthquake.usgs.gov/fdsnws/event/1/count"

# Let's use variables for three of the parameters:
min_mag = 5.0               # the minimum magnitude considered a quake (min_mag)
start_time = "2025-01-01"   # this is the year-month-day format of the start
finish_time = "2025-02-01"  # similar for the end

# we assemble a dictionary of our parameters, let's name it param_dictionary
# there are many more parameters available. The problems below ask you to explore them...
param_dictionary = { "format":"geojson",         # this is simply hard-coded to obtain json
                     "starttime":start_time,
                     "endtime":finish_time,
                     "minmagnitude":min_mag,
                     }

# Here, we use requests to make the request. The parameters will be added by this API call:
result = requests.get(url, params=param_dictionary)
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # this will include the parameters!

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&starttime=2025-01-01&endtime=2025-02-01&minmagnitude=5.0


In [8]:
# If it worked, we should be able to see the json results:

d = result.json()
print(f"JSON returned was {d = }")

JSON returned was d = {'count': 149, 'maxAllowed': 20000}


From here, it would be possible to write one or more loops and build an earthquake dataset. For example,
+ it would be possible to loop over the ``minmagnitude`` to get a distribution of different sized quakes (or a histogram)
+ it would be possible to loop over one of the <i>time-endpoints</i>
+ it would be possible to loop over one of the _other parameters_ e.g.,
  + you can specify a circle around a specific ``latitude`` and ``longitude`` with a ``maxradiuskm`` (the radius)
  + Claremont is at ``latitude=34.0967`` and ``longitude=-117.7198`` 

The next two cells have an example of a Claremont-centric quake-count question and answer:

In [11]:
#
# How many quakes of magnitude >= 4.2 have been within 300km of Claremont 
#     + in Jan 2025
#     + in Dec 2025
#
url = "https://earthquake.usgs.gov/fdsnws/event/1/count"

# Let's use variables for three of the parameters:
min_mag = 2.2               # the minimum magnitude considered a quake (min_mag)
start_time = "2025-01-01"   # this is the year-month-day format of the start
finish_time = "2025-02-01"  # similar for the end
# start_time = "2024-01-01"   # similar, but for a year-CS35_Participant_2 span...
# finish_time = "2025-01-01"  # similar for the end
radius_in_km = 300

# we assemble a dictionary of our parameters, let's name it param_dictionary
# there are many more parameters available. The problems below ask you to explore them...
param_dictionary = { "format":"geojson",         # this is simply hard-coded to obtain json
                     "starttime":start_time,
                     "endtime":finish_time,
                     "minmagnitude":min_mag,
                     "latitude":34.0967,
                     "longitude":-117.7198,
                     "maxradiuskm":radius_in_km,
                     }

# Here, we use requests to make the request. The parameters will be added by this API call:
result = requests.get(url, params=param_dictionary)
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # this will include the parameters!

# We'll extract the final result in another cell:

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&starttime=2025-01-01&endtime=2025-02-01&minmagnitude=2.2&latitude=34.0967&longitude=-117.7198&maxradiuskm=300


In [ ]:
# Let's finish up here:
quake_count = result.json()
print(f"{quake_count = }")

quake_count = {'count': 70, 'maxAllowed': 20000}


<font color="Coral"><b>Try this, task "0b"</b></font>  &nbsp;&nbsp; 
+ Would you expect more or fewer quakes if a minimum magnitude of 2.2 were used? **More! The plates beneath us are constantly shifting and we might not feel it.**
+ Try the above cells again for a minimum magnitude of 2.2 -- if that the difference you'd expect? **Yes count went from 1 to 70.**

<font color="DodgerBlue"><b>Quake-counting Results:</b></font> 

#### Number of Claremont-centric quakes
  + <u>Overall</u> The API calls to the USGS showed that, within 300km of Claremont, there were
    + 1 quake of magnitude 4.2 or larger within 300km of Claremont in Jan '25
    + more quakes  of magitude 2.2 or larger within 300km of Claremont in Jan '25 <br><br>
  + <u>Reflection</u>: _This is not enough data to establish a trend. (I hope!)_ That said, for the hw, you should include a **loop** over at least 10 values, as well as a text-formatted set of values. (It's ok for the output values to be in the computational cell(s), not the markdown cell. The markdown is really for reflection on results than reporting of results.) <br><br>
  + <u>Opportunities</u>: In fact, there are lots of values over which the USGS API allows to vary. For example,  minmagnitude (or maxmagnitude), radius, location (lat/CS35_Participant_2), months (or other time-measurements) - and others. In addition, there is the chance to look at the details of each quake using the ``query`` endpoint. <br><br>

Looping over API calls to gather data will be one of the hw problems.

<hr>

#### The ``query`` endpoint

Let's see, too, that it's possible to obtain, not only a <i>count</i>, but also a <b>"full report"</b> of all of the earthquakes.

To do so, the only change needed is from the endpoint ``count`` to the endpoint ``query``

First, try it "by hand" -- by opening this url in your browser:

[https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&minmagnitude=6.8&starttime=2024-01-01&endtime=2024-02-01](https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&minmagnitude=6.8&starttime=2024-01-01&endtime=2024-02-01)

Take a look -- there is a lot more data!  

Next, let's try it programmatically:

In [28]:
#
# Here is the endpoint
#
url = "https://earthquake.usgs.gov/fdsnws/event/1/query"

# Let's use variables for three of the parameters:
min_mag = 6.5 
max_mag = 7.0              # the minimum magnitude considered a quake (min_mag)
start_time = "2025-01-01"   # this is the year-month-day format of the start
finish_time = "2025-02-01"  # similar for the end

# we assemble a dictionary of our parameters, let's name it param_dictionary
# there are many more parameters available. The problems below ask you to explore them...
param_dictionary = { "format":"geojson",         # this is simply hard-coded to obtain json
                     "starttime":start_time,
                     "endtime":finish_time,
                     "minmagnitude":min_mag,
                     "maxmagnitude":max_mag
                     }

# Here, we use requests to make the request. The parameters will be added by this API call:
result = requests.get(url, params=param_dictionary)
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # this will include the parameters!

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2025-01-01&endtime=2025-02-01&minmagnitude=6.5&maxmagnitude=7.0


In [ ]:
# If it worked, we should be able to see the json results:

d = result.json()
print(f"JSON returned was {d = }")

In [ ]:
#
# That's hard to read!
# Let's pretty-print it with the json library
#       Also, this version can be pasted into online formatters, e.g., https://jsonformatter.org/

import json 
nice_string = json.dumps(d)   # this outputs a "nicely formatted string" using double quotes
print(nice_string)


but, we can do better! 

the "dump string" function, json.dumps, can output the formatted version, too...


In [ ]:
import json 
nicer_string = json.dumps(d, indent=4)   # We can specify the indentation. 
print(nicer_string)                      # It's another tree structure... !

<font color="Coral"><b>Try this, also for Task 0b</b></font>  &nbsp;&nbsp; 
+ Look through the response to find where and when the largest eathquake was, in the previous year. **Southern Tibetan Plateau **
+ What was its magnitude? **7.1**
+ (optional) Do you see a way that - just by making small changes and re-running - you could find the second-biggest quake of '24?  Perhaps try it...  🦔 **added maximum magnitude into the dictionary and defined max_mag**

<br>
<hr>
<br>



#### Launching into hw2's challenges:
+ <font color="Coral"><b>Tasks1-2</b></font> &nbsp;&nbsp; ISS challenges:
  + ``ISS_now()`` which will find and return the ISS's lat/CS35_Participant_2 (as floats)
    + use the earlier examples as a starting point...
  + ``ISS_distance()`` which will return the distance of the ISS from a city of your choice (a constant city, not a variable). It can be Claremont, but doesn't have to be. This will require using the <i>haversine</i> distance for global lat/CS35_Participant_2 coordinates...
+ <font color="Coral"><b>Tasks3-4</b></font> &nbsp;&nbsp; Earthquake challenges:
  + ``Quake_loop()`` which will loop over a parameter of your choice, print a formatted list of quake-count data, and return that list
  + ``Quake_compare(place1, place2)`` which will ask - and answer - a comparative question about "quakiness" for two different places, using the quake data... 
    + Here, the goal is to define which of the two places is "quakiest."
    + Notice that the definition of "quakiest" is up to you...
    + The inputs, place1 and place2, can be lat/CS35_Participant_2 pairs - or strings, which then get looked up...

<br>

#### <font color="Coral"><b>Task5</b></font> &nbsp;&nbsp; Open-ended, two-hop challenge

Then, you'll choose or create an open-ended API task -- or create a variant of your own design from at least two APIs - or "scrapes" - of your choice.
+ The key constraint is to be sure you meaningfully use the  <font color="DodgerBlue">result</font>  from the first API call in order to customize the request of the second API call.
+ Then, with the result of that second API call, interpret the data obtained to make a final "statement" -- which can be anywhere amid serious, silly, surreal -- that combines the two API insights.
+ See below for a far-fetched superbowl-themed idea...




In [49]:
#
# hw2: ISS tasks 1 and 2 ...
# 
# Two functions:  ISS_now(), ISS_distance()
import requests

url = "http://api.open-notify.org/iss-now.json"   # this is sometimes called an "endpoint" ...
result = requests.get(url)
json_contents = result.json()

#
# Use the ISS examples above to write a function, named 
#     
#      ISS_now()
#
# that uses requests to return the current latitude and longitude -- as floating-point values -- right now.
# Be sure to test it! 

def ISS_now():
    lat1 = float(json_contents['iss_position']['latitude'])
    long1 = float(json_contents['iss_position']['longitude'])
    return lat1, long1

ISS_now()


(-45.5388, -131.7348)

Feel free to create some new cells around this area to write and test ``ISS_now()`` ...

In [52]:
# 
# Once your ISS_now() function is working, write a new function
#
#       ISS_distance()
#
# which uses ISS_now to obtain the lat/CS35_Participant_2 of the ISS and then
# uses the haversine distance (look up a Python implementation or use one of ours... :)
# to compute the ISS's distance from a city of your choice.
#
# The haversine distance computes the "great circle" distance from two points on the globe
#     using latitude and longitude  
#
def ISS_distance():
    """
    Calculate the distance of the ISS at any given time to Manila, Philippines
    """
    from math import sin, cos, radians, asin, sqrt   # this import is for the sin, cos, radians, ...
    # convert decimal degrees to radians

    lat1, long1 = ISS_now()
    # Hardcoded the latitude and longitude of Manila, Philippines
    lat2 = 14.5995
    long2 = 120.9842

    long1, lat1, long2, lat2 = map(radians, [long1, lat1, long2, lat2])

    # haversine formula
    dlong = long2 - long1
    dlat = lat2 - lat1
    trig = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlong/2)**2
    
    # Radius of earth. Use 3956 for miles. 6371 for km.
    radius = 3956  # we'll use miles!
    return radius * 2 * asin(sqrt(trig))

distance = ISS_distance()

# Print the result
print(f"Distance of ISS from Manila, PH: {distance:.2f} miles")


Distance of ISS from Manila, PH: 7761.46 miles


Feel free to create some new cells around this area to write and test ``ISS_distance()`` ...

In [ ]:
#
# Open-ended possibility:  
#    (a) create a new function ISS_distance(place) that takes in a place name
#    (b) find a service by which you can look up the lat + CS35_Participant_2 using the place name
#         (b*)  I'm not sure how to do this - it's exploratory! 
#    (c) then, continue with the previous computation to find the ISS distance! :) 
#
d = {}

import requests
from math import sin, cos, radians, asin, sqrt

def ISS_distance(place):

    url = f"https://api.geodatasource.com/format=json&"
    result = requests.get(url).json()
    
    if result:
        lat = d[place]
        CS35_Participant_2 = float(result[0]['CS35_Participant_2'])
        return lat, CS35_Participant_2
    
place_name = "Manila"
distance = ISS_distance(place_name)

print(f"Distance of ISS from {place_name}: {distance:.2f} miles")

# The final problem of this hw2 is to take on _ONE_ open-ended possibility. 
#     (This ISS-themed one is only the first possibility.)
#     Others, below, involve earthquakes, or your own choice of API exploration...

#### USGS Challenges

Tasks 3 and 4 use the earthquake API

<font color="Coral"><b>Tasks3-4</b></font> &nbsp;&nbsp; Earthquake challenges:
  + ``Quake_loop()`` which will loop over a parameter of your choice, print a formatted list of quake-count data, and return that list
  + ``Quake_compare(place1, place2)`` which will ask - and answer - a comparative question about "quakiness" for two different places, using the quake data... 
    + Here, the goal is to define which of the two places is "quakiest."
    + Notice that the definition of "quakiest" is up to you...
    + The inputs, place1 and place2, can be lat/CS35_Participant_2 pairs - or strings, which then get looked up...

In [204]:
import requests
import time

def get_num_quakes(start_time, finish_time, min_mag, latitude, longitude, max_radius_km):
   """ returns the number of quakes in month (of '24)
           of at least magnitude ...
   """
   url = "https://earthquake.usgs.gov/fdsnws/event/1/count"
 
   # we assemble a dictionary of our parameters, let's name it param_dictionary
   # there are many more parameters available. The problems below ask you to explore them...
   param_dictionary = { "format":"geojson",         # this is simply hard-coded to obtain json
                        "starttime":start_time,
                        "endtime":finish_time,
                        "minmagnitude":min_mag,
                        "latitude": latitude,         
                        "longitude": longitude,      
                        "maxradiuskm": max_radius_km
                       }
   
   result = requests.get(url, params=param_dictionary)
   # return f"({result} the result.url is {result.url})"
   d = result.json()
   count = d.get('count')
   return count

# Manila to rest of Luzon is a 200 km radius but max_radius_km max is 180
print(get_num_quakes("2024-01-01","2024-04-01",4.0,14.5995,120.9842,180))
print(get_num_quakes("2024-01-01","2024-04-01",4.0,34.0967,-117.7198,180))

10
4


In [206]:
#
# hw2: USGS Tasks 3 and 4 ...
# 
# Two functions:  Quake_loop(), Quake_compare(place1, place2)

#
# Use the USGS (earthquake) examples above to write a function, named 
#     
import requests
import time

magnitudes = [2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0] # the strongest earthquake magnitude was 9.5 in 1960!
LoQ = []

def Quake_loop():
    ''' loops over different magnitudes to collect earthquake counts ''' 
    start_time = "2024-01-01"   
    finish_time = "2024-04-01"
    
    latitude = 14.5995 #34.0967 
    longitude = 120.9842 #-117.7198 
    max_radius_km = 180  

    LoQ = []

    for magnitude in magnitudes:
        count = get_num_quakes(start_time, finish_time, magnitude, latitude, longitude, max_radius_km)
        LoQ.append(count)
        
    return LoQ
LoQ = Quake_loop()

print(f"Between 2024-01-01 and 2024-04-01, there were the following earthquakes")
for i, magnitude in enumerate([2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]):
    print(f"Magnitude {magnitude}: {LoQ[i]} quakes")
#
# that uses requests within a loop of your own design in order to
#   + obtain at least 10 distinct, comparable data elements (counts are encouraged; other items ok)
#   + see the assignment page for an example where the looping iterates over the _month_
#
#   + choose your favorite parameter(s) to vary, e.g., magnitude, time, radius, location, etc.
#   + it should collect all of those data elements into a list
#   + and render the list in a neatly formatted chart (f-strings welcome; not required)
#
#   + in addition, include a overall reflection on the results, as well as a comment on additional effort
#     that could expand your results (you need not run it), and any concerns or caveats about the data...
#   + feel free to copy-paste-edit the markdown "reflection-template," above  

Between 2024-01-01 and 2024-04-01, there were the following earthquakes
Magnitude 2.0: 10 quakes
Magnitude 3.0: 10 quakes
Magnitude 4.0: 10 quakes
Magnitude 5.0: 0 quakes
Magnitude 6.0: 0 quakes
Magnitude 7.0: 0 quakes
Magnitude 8.0: 0 quakes
Magnitude 9.0: 0 quakes
Magnitude 10.0: 0 quakes


#### Number of Quakes within Luzon
  + Using Manila as the reference point, 180 km radius captures most of the Luzon region. 
  + There are many minor and undetected quakes - 17k! The earth is constantly shifting beneath our feet. 
  + I wonder where the two 7.0 quakes were it would be more informative if the function could pinpoint the locations of the larger earthquakes, defined by a certain magnitude threshold. 

Feel free to create some new cells around this area to write and test ``Quake_loop()`` ...

In [208]:
# 
# Once your Quake_loop() function is working, write a new function
#
def Quake_compare(place1, place2):
    ''' determines which of two places is quakier based on specified time span, magnitude, and radius around the two chosen places'''
#
# and then your function should compare which of the two places is "quakier" (not a real word)
# for a given time span (you choose), and a given strength-of-quakes (you choose), and
# for a specific radius around each of the two places (you choose)
#
    start_time = f"2024-01-01"   
    finish_time = f"2024-04-01"
    min_mag = 4.0
    max_radius_km = 180  

    lat1, long1 = place1 
    lat2, long2 = place2

    count1 = get_num_quakes(start_time, finish_time, min_mag, lat1, long1, max_radius_km)
    count2 = get_num_quakes(start_time, finish_time, min_mag, lat2, long2, max_radius_km)

    print(f"Between {start_time} and {finish_time}, with magnitude ≥ {min_mag} and radius ≤ {max_radius_km} km:")
    print(f"Claremont {place1} had {count1} quakes.")
    print(f"Manila {place2} had {count2} quakes.")

    if count1 > count2:
        print(f"Claremont {place1} is quakier than Manila {place2}.")
    elif count2 > count1:
        print(f"Manila {place2} is quakier than Claremont {place1}.")
    else:
        print(f"Claremont {place1} and Manila {place2} are just as quaky.")

place1 = (34.0967,-117.7198) # Claremont
place2 = (14.5995,120.9842) # Manila

Quake_compare(place1,place2)

# As is clear, there is lots of freedom to design a "comparison of quakiness" -- wonderful!
# Feel free to start somewhere, and tune the results.
#
# Your function should print the results it finds (in this case, it's not important to return
# and specific value -- though you're encouraged (not required) to create a helper function and 
# then call it twice for the two locations! (That helper function would need a return value!)
#
#

Between 2024-01-01 and 2024-04-01, with magnitude ≥ 4.0 and radius ≤ 180 km:
Claremont (34.0967, -117.7198) had 4 quakes.
Manila (14.5995, 120.9842) had 10 quakes.
Manila (14.5995, 120.9842) is quakier than Claremont (34.0967, -117.7198).


Feel free to create some new cells around this area to write and test ``Quake_compare(place1, place2)`` ...

#### Final API challenge:  an open-ended, two-hop API task...

<font color="Coral">Key constraint: _Use two hops!_</font>  &nbsp;&nbsp; Be sure to dome something that uses two separate API calls, in which the results of the first affect the second - culminating in an aggregate, overall result.

Possibilities: 
  + You should use at least one other API. (See the hw2 page for links to many.)
      + The [Poke API](https://pokeapi.co/)  or [one of these](https://medium.com/codex/15-fun-and-interesting-apis-to-use-for-your-next-coding-project-in-2022-86a4ff3a2742) or [one of the many, many more!](https://github.com/public-apis/public-apis)
  + That said, <u><b>one</b></u> of the two can be ISS or USGS, as you used above.



<font size="-2">PS. My example of this is in ``superbowl_prediction.ipynb`` with the starter files. It used to be here, but it was too clutter-y... .</font>

<br><hr><br>

Big-picture, an important part of using API calls is gathering a _specific piece_ of data from an otherwise too-large ocean of raw material. That element then allows you to express a natural _next_ specific question, obtain its relevant data, and continue to build from there, with each round-trip branching into additional (possible) questions... .

<br>

Good luck API'ing!!

<br>



In [227]:
#
import requests

# 

def get_ye_quotes():
   """ returns a funfact about the number of letters in a random kanye west quote ...
   """
   url = "https://api.kanye.rest"

   result = requests.get(url)
   # return f"({result} the result.url is {result.url})"
   d = result.json()
   quote = d.get('quote')
   print(quote)
   number = str(len(quote)) 
   
   url2 = f"http://numbersapi.com/{number}/trivia"

   result2 = requests.get(url2)
   print(result2.text) # get plain text

get_ye_quotes()

#

I love UZI. I be saying the same thing about Steve Jobs. I be feeling just like UZI
83 is the highest UHF channel on older televisions made before the late 1970s.
